File: security_agency/security_agency/page/gps_checkin_dashboar/gps_checkin_dashboar.js
frappe.pages['gps-checkin-dashboar'].on_page_load = function(wrapper) {

    console.log("🚀 gps-checkin-dashboar page JS loaded");

    const page = frappe.ui.make_app_page({
        parent: wrapper,
        title: 'GPS Check-in Dashboard',
        single_column: true
    });

    // Render the HTML template
    $(frappe.render_template("gps_checkin_dashboar", {})).appendTo(page.body);

    let chartSite = null;
    let chartLast7 = null;

    // Load Chart.js
    $.getScript("https://cdn.jsdelivr.net/npm/chart.js", function () {
        load_site_filter();
        set_default_date();
        load_dashboard();
    });

    // 🔥 Load Site dropdown from backend
    function load_site_filter() {
        frappe.call({
            method: "security_agency.security_agency.page.gps_checkin_dashboar.gps_checkin_dashboar.get_sites",
            callback: function (r) {
                let sites = r.message || [];
                let select = $("#filter_site");

                sites.forEach(s => {
                    select.append(`<option value="${s.name}">${s.name}</option>`);
                });
            }
        });

        // Bind filter button
        $("#apply_filters").click(function () {
            load_dashboard();
        });
    }

    // Set default date = today
    function set_default_date() {
        $("#filter_date").val(frappe.datetime.get_today());
    }

    // 🔥 Load Dashboard Data
    function load_dashboard() {

        const site = $("#filter_site").val();
        const date = $("#filter_date").val();

        console.log("📡 Sending filters:", site, date);

        frappe.call({
            method: "security_agency.security_agency.page.gps_checkin_dashboar.gps_checkin_dashboar.get_data",
            args: {
                site: site,
                date: date
            },
            callback: function (r) {
                const d = r.message || {};
                console.log("📥 Dashboard Data:", d);

                $('#total_today').text(d.total_today || 0);
                $('#pending').text(d.pending || 0);
                $('#approved').text(d.approved || 0);
                $('#rejected').text(d.rejected || 0);

                // Destroy existing charts before creating new ones
                if (chartSite) chartSite.destroy();
                if (chartLast7) chartLast7.destroy();

                // 🔵 Chart 1 — By Site
                chartSite = new Chart(document.getElementById('chart_by_site').getContext("2d"), {
                    type: 'bar',
                    data: {
                        labels: (d.by_site || []).map(x => x.site),
                        datasets: [{
                            label: 'Check-ins',
                            data: (d.by_site || []).map(x => x.count),
                            backgroundColor: 'rgba(54,162,235,0.6)'
                        }]
                    }
                });

                // 🟢 Chart 2 — Last 7 Days
                chartLast7 = new Chart(document.getElementById('chart_last7').getContext("2d"), {
                    type: 'line',
                    data: {
                        labels: (d.last7 || []).map(x => x.date),
                        datasets: [{
                            label: 'Check-ins',
                            data: (d.last7 || []).map(x => x.count),
                            borderColor: 'rgba(75,192,192,1)',
                            fill: false,
                            tension: 0.1
                        }]
                    }
                });
            }
        });
    }
};

File: security_agency/security_agency/page/gps_checkin_dashboar/gps_checkin_dashboar.py
import frappe
from frappe.utils import today, add_days

@frappe.whitelist()
def get_data(site=None, date=None):
    """
    Dashboard numbers and chart data for GPS Check-in Request.
    Now supports filters:
    - site (optional)
    - date (optional, default: today)
    """

    # --- Date Filter ---
    selected_date = date or today()
    day_start = selected_date + " 00:00:00"
    day_end = selected_date + " 23:59:59"

    # Last 7 days from selected date
    week_ago = add_days(selected_date, -6)

    # --- Dynamic WHERE conditions ---
    where_site = ""
    params_site = []

    if site:
        where_site = "AND g.site = %s"
        params_site.append(site)

    # --- Total Check-ins ---
    total_today = frappe.db.sql(
        f"""
        SELECT COUNT(*)
        FROM `tabGPS Check-in Request` g
        WHERE g.check_in_time BETWEEN %s AND %s
        {where_site}
        """,
        [day_start, day_end] + params_site
    )[0][0]

    # --- Validated Status Count ---
    def validated_count(pattern):
        return frappe.db.sql(
            f"""
            SELECT COUNT(*)
            FROM `tabGPS Check-in Request` g
            WHERE g.workflow_state LIKE %s
              AND g.check_in_time BETWEEN %s AND %s
              {where_site}
              AND EXISTS (
                  SELECT 1
                  FROM `tabDeployment Sheet` s
                  INNER JOIN `tabDeployment Line` l ON l.parent = s.name
                  WHERE g.employee = l.guard
                    AND g.site = l.site
                    AND l.day = DAYNAME(g.check_in_time)
                    AND DATE(g.check_in_time) BETWEEN s.week_start_date AND s.week_end_date
              )
            """,
            [pattern, day_start, day_end] + params_site
        )[0][0]

    pending = validated_count("Pending%")
    approved = validated_count("Approved%")
    rejected = validated_count("Rejected%")

    # --- Chart 1: By Site ---
    by_site = frappe.db.sql(
        f"""
        SELECT IFNULL(site, 'No Site') AS site, COUNT(*) AS count
        FROM `tabGPS Check-in Request` g
        WHERE g.check_in_time BETWEEN %s AND %s
        {where_site}
        GROUP BY IFNULL(site, 'No Site')
        """,
        [week_ago, day_end] + params_site,
        as_dict=True
    )

    # --- Chart 2: Last 7 Days ---
    last7 = frappe.db.sql(
        f"""
        SELECT DATE(g.check_in_time) AS date, COUNT(*) AS count
        FROM `tabGPS Check-in Request` g
        WHERE g.check_in_time BETWEEN %s AND %s
        {where_site}
        GROUP BY DATE(g.check_in_time)
        ORDER BY date
        """,
        [week_ago, day_end] + params_site,
        as_dict=True
    )

    # Debug log
    frappe.logger().info({
        "filters": {"site": site, "date": date},
        "by_site": by_site,
        "last7": last7
    })

    return {
        "total_today": total_today,
        "pending": pending,
        "approved": approved,
        "rejected": rejected,
        "by_site": by_site,
        "last7": last7
    }


@frappe.whitelist()
def get_sites():
    """Return list of sites for dropdown."""
    return frappe.db.get_all("Site", fields=["name"])
